[quant][fx] produce reference pattern for binary ops and then rewrite to quantized op

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73386
* #73385
* #73233
* #73120
* __->__ #72953

Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we have
some custom lowering pass to rewrite the reference qunatized patterns to quantized ops
it includes rewrite for
torch.ops.quantized.add, torch.ops.quantized.mul, torch.ops.quantized.matmul

Test Plan:

python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34292408](https://our.internmc.facebook.com/intern/diff/D34292408)

jerryzh168 committed:
[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op

Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we have
some custom lowering pass to rewrite the reference qunatized patterns to quantized ops
it includes rewrite for
torch.ops.quantized.add, torch.ops.quantized.mul, torch.ops.quantized.matmul

Test Plan:

python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we have
some custom lowering pass to rewrite the reference qunatized patterns to quantized ops
it includes rewrite for
torch.ops.quantized.add, torch.ops.quantized.mul, torch.ops.quantized.matmul

Test Plan:

python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34292408).

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we have
some custom lowering pass to rewrite the reference qunatized patterns to quantized ops
it includes rewrite for
torch.ops.quantized.add, torch.ops.quantized.mul, torch.ops.quantized.matmul

Test Plan:

python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34292408](https://our.internmc.facebook.com/intern/diff/D34292408)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we have
some custom lowering pass to rewrite the reference qunatized patterns to quantized ops
it includes rewrite for
torch.ops.quantized.add, torch.ops.quantized.mul, torch.ops.quantized.matmul

Test Plan:

python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34292408](https://our.internmc.facebook.com/intern/diff/D34292408)

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/match_utils.py:
```diff
@@ -47,24 +47,26 @@ def is_match(modules, node, pattern, max_uses=sys.maxsize):
         return True
 
-    if len(node.users) > max_uses:
+    if isinstance(node, Node) and len(node.users) > max_uses:
```
why is this check necessary, shouldn't this always be a node?

jerryzh168 replied:
oh this is added during debugging. theoretically it could be a constant as well I think, but I think we can add that later. will remove the changes to this file if it doesn't break

vkuzo commented on torch/ao/quantization/fx/quantization_patterns.py:
```diff
@@ -523,7 +409,28 @@ def modified_load_arg(n: Node):
+            kwargs = load_arg(quantized=torch.float)(self.binary_op_node.kwargs)
+            op_out = quantized_graph.node_copy(self.binary_op_node, load_arg(quantized=torch.float))
+
+            def modified_load_arg(n: Node):
```
I realize this was in the original code, but would be great to remove this in the future to increase readability

jerryzh168 replied:
yeah 90% of code here will be removed after the refactor, including this and the overall load_arg stuff

andrewor14 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -130,9 +137,149 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule) -> QuantizedGraphMod
+             MatchAllNode, MatchAllNode, MatchAllNode)
+            for relu_op in [torch.relu, torch.nn.functional.relu, torch.nn.ReLU]
+            for bop_pattern in [
+                (bop, "dequantize", "dequantize"),
```
Do we need `(bop, "dequantize", "dequantize")`, since anything that matches this one would also match the next two patterns?

andrewor14 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -130,9 +137,149 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule) -> QuantizedGraphMod
+            (torch.quantize_per_tensor,
+             (bop, "dequantize", "dequantize"),
+             MatchAllNode, MatchAllNode, MatchAllNode)
+            for bop_pattern in [
```
This is not used. Did you mean the following?
```
(torch.quantize_per_tensor, bop_pattern, MatchAllNode, MatchAllNode, MatchAllNode)
```

jerryzh168 replied:
ah sorry, let me fix it

andrewor14 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -130,9 +137,149 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule) -> QuantizedGraphMod
+            ]
+        ]
+
+    def get_bop_relu_patterns(bop: Callable) -> List[Pattern]:
```
nit: I wonder if there's a way to merge these two and still be clear, maybe like:
```
def get_bop_patterns(bop):
    patterns = []
    for bop_pattern in [...]:
        patterns.append(
            (torch.quantize_per_tensor,
                bop_pattern,
                MatchAllNode, MatchAllNode, MatchAllNode))
        for relu_op in [...]:
            patterns.append(
                (torch.quantize_per_tensor,
                     (relu_op, bop_pattern),
                     MatchAllNode, MatchAllNode, MatchAllNode))
    return patterns
```
What do you think?

jerryzh168 replied:
makes sense

andrewor14 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -130,9 +137,149 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule) -> QuantizedGraphMod
+            for bop_pattern in [
+                (bop, "dequantize", "dequantize"),
+                (bop, MatchAllNode, "dequantize"),
+                (bop, "dequantize", MatchAllNode),
```
Do we need (bop, "dequantize", "dequantize"), since anything that matches this one would also match these two patterns?

jerryzh168 replied:
god point, this is referring to add/add_scalar, maybe we could just use MatchAllNode for both arguments, and filter the things we want later, this would reduce the number of patterns.

jerryzh168 committed:
Update on "[quant][graphmode] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we have
some custom lowering pass to rewrite the reference qunatized patterns to quantized ops
it includes rewrite for
torch.ops.quantized.add, torch.ops.quantized.mul, torch.ops.quantized.matmul

Test Plan:

python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34292408](https://our.internmc.facebook.com/intern/diff/D34292408)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34292408).

jerryzh168 committed:
Update on "[quant][fx] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we have
some custom lowering pass to rewrite the reference qunatized patterns to quantized ops
it includes rewrite for
torch.ops.quantized.add, torch.ops.quantized.mul, torch.ops.quantized.matmul

Test Plan:

python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34292408](https://our.internmc.facebook.com/intern/diff/D34292408)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34292408).

jerryzh168 committed:
Update on "[quant][fx] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we have
some custom lowering pass to rewrite the reference qunatized patterns to quantized ops
it includes rewrite for
torch.ops.quantized.add, torch.ops.quantized.mul, torch.ops.quantized.matmul

Test Plan:

python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34292408](https://our.internmc.facebook.com/intern/diff/D34292408)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we have
some custom lowering pass to rewrite the reference qunatized patterns to quantized ops
it includes rewrite for
torch.ops.quantized.add, torch.ops.quantized.mul, torch.ops.quantized.matmul

Test Plan:

python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34292408](https://our.internmc.facebook.com/intern/diff/D34292408)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] produce reference pattern for binary ops and then rewrite to quantized op"


Summary:
This PR makes BinaryOpQuantizeHandler to always produce reference patterns, and we have
some custom lowering pass to rewrite the reference qunatized patterns to quantized ops
it includes rewrite for
torch.ops.quantized.add, torch.ops.quantized.mul, torch.ops.quantized.matmul

Test Plan:

python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34292408](https://our.internmc.facebook.com/intern/diff/D34292408)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34292408).
